### PR TITLE
fix(import): Logseq import — create pages, preserve content, resolve backlinks (#43)

### LINKED ISSUE #43: Logseq import (`outl import logseq`) creates blocks but no pages — `page list` empty, backlinks broken
## Summary

Bulk-ingesting existing `.md` files never creates the **page** itself — only its blocks. After `outl import logseq`, `outl serve --once`, or simply copying `.md` files into `pages/`/`journals/` and opening the mobile/TUI client, `outl page list` returns `no pages`, backlinks resolve to nothing, and the imported content is effectively invisible even though every block lands in the op log.

This breaks the primary onboarding path for Roam/Logseq refugees, which is exactly the audience the project targets.

## Reproduce

```sh
outl import logseq <logseq-graph> /tmp/ws
outl page list -w /tmp/ws          # => "no pages"
grep -c page-slug /tmp/ws/ops/*.jsonl   # => 0
```

Same result with `outl serve --once` over a directory of clean outl-dialect `.md` files, and with the mobile/TUI orphan scanner on startup.

For comparison, the only path that works:

```sh
outl page create foo --title Foo -w /tmp/ws
outl page list -w /tmp/ws          # => "page  foo  Foo"
grep -o '"key":"page-[a-z]*"' /tmp/ws/ops/*.jsonl   # => page-slug, page-kind
```

## Expected vs actual

- **Expected:** ingesting `pages/foo.md` makes `foo` show up as a page (with backlinks working).
- **Actual:** the blocks exist in the op log but hang off a `page_id` that is never created under `root` and never tagged with `page-slug`, so `page::list_all` (which filters children of root by `page-slug`) finds nothing.

## Root cause

The page entity is created **only** by `outl_actions::page::open_or_create`, which emits `Op::Create` under `root` + `SetProp page-slug` + `SetProp page-kind` (`crates/outl-actions/src/page.rs:185`). The bulk-ingest path never calls it:

- `reconcile_md` takes a `page_id` and, when no sidecar exists, mints a **random** `NodeId::new()` for it (`crates/outl-md/src/reconcile.rs:78-87`).
- `diff_to_ops` receives that `page_id` and only ever emits `Op::Create` for the **child blocks** with `parent = page_id`; it never creates the page node itself nor sets `page-slug` (`crates/outl-md/src/diff.rs:97`, `:167`, `:179`).
- `outl import` → `seed_sidecars` only loops `reconcile_md` over each file (`crates/outl-cli/src/cmd/import.rs:235`) — no `open_or_create`.
- Mobile startup `reconcile_orphan_md` and the TUI orphan scanner likewise only call `reconcile_md` (`crates/outl-mobile/src-tauri/src/lib.rs:645-656`, `crates/outl-tui/src/actions/lifecycle.rs:258-301`).

There is no helper that combines "ingest a `.md` as a page" (open_or_create + seed sidecar at the deterministic id + reconcile).

Secondary mismatch: `reconcile_md` uses a random `page_id` for new sidecars, while `open_or_create` uses the deterministic `page_id_from_slug`. Even if a page node were created first, the two ids would diverge and the blocks would attach to a phantom node.

## Impact

- `outl import logseq|roam` produces a workspace with zero visible pages (the documented onboarding flow).
- Copying `.md` into an iCloud-synced workspace and opening mobile/TUI shows nothing — the scenario that surfaced this.
- Backlinks never resolve for ingested content (they key off pages that don't exist as page nodes).

## Proposed fix

Make bulk ingest create the page entity. Options, roughly in order of preference:

1. Add a `reconcile_md_as_page` (or a flag on `reconcile_md`) in `outl-md`/`outl-actions` that, for a file under `pages/`/`journals/`:
   - derives `slug` from the filename and `kind` from the directory,
   - reads `title::` (falling back to slug),
   - calls `open_or_create` with the deterministic `page_id_from_slug(slug)`,
   - then reconciles blocks under that page node.
2. Route `import::seed_sidecars`, `outl serve` initial scan, and the mobile/TUI orphan scanners through that helper instead of bare `reconcile_md`.
3. Align `reconcile_md`'s new-sidecar `page_id` with `page_id_from_slug(slug)` so independent devices converge on the same page node.

Acceptance: after `outl import logseq <graph> <ws>`, `outl page list -w <ws>` lists every imported page and `outl backlinks page <slug>` resolves `[[refs]]`.

## Notes / context

Surfaced while restoring a backup produced by the old SQLite-log era (0.4.x). `outl migrate-to-shared` correctly copied all 13,777 ops from `.outl/log.db` into `ops/ops-<actor>.jsonl` (op-count parity), and `outl doctor` reports "all IDs known" for every sidecar — so the tree migrated fine. The pages were still invisible purely because of this ingest gap (the pre-page-model 0.4.x ops carry no `page-slug` either).

### PR DESCRIPTION
**Pull request by vmesel**
fix(import): make Logseq import preserve pages, content, and backlinks

Bulk `.md` ingest (import, serve, mobile/TUI orphan scanners) created
blocks but never the page node with `page-slug`, so imported graphs
showed no pages and backlinks resolved to nothing (#43). Validating the
fix on a real Logseq graph surfaced more import gaps that lost content
silently.

Page creation on ingest:
- Add `outl_actions::ingest_md_file`/`ingest_dir`: create the page node
  (open_or_create, deterministic id via page_id_from_slug) then
  reconcile blocks under it.
- Add `outl_md::reconcile_md_with_page_id` to pin the page id when no
  sidecar exists yet, so blocks attach to the page open_or_create made.
- Route serve (sync_engine), import (seed_sidecars), and the mobile/TUI
  orphan scanners through ingest.

Implicit pages (Logseq parity):
- Add `outl_actions::create_missing_ref_pages`: create a stub page for
  every `[[ref]]` without its own file (title = the reference text so
  backlinks match), date refs become journals. Each stub is projected
  to a `.md` + sidecar so the disk-scanning TUI lists and can open it,
  not just the op-log clients. Called at end of import.

Outline normalization (no silent content loss):
- New import/normalize.rs rebuilds free-form Logseq markdown into the
  strict outl outliner dialect: loose `## headings`/paragraphs become
  bullets, multi-paragraph block bodies merge (blank lines dropped),
  orphan indented blocks get their level clamped, and code fences are
  dedented + re-indented preserving the code's relative indentation.
- import write_page_md no longer prepends a second `title::` when the
  body already carries page properties.

Result on a 369-file Logseq graph: every page with 100% of content
preserved (was 0 visible pages), references resolve to a page,
backlinks working, code fences intact.

Closes #43

### CODE REVIEW

File: crates/outl-cli/src/cmd/import/normalize.rs [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+/// alignment, keeps the code's relative structure), and append the
+/// dedented body plus a bare closing fence to `text`. Re-indentation to
+/// the block's canonical level happens at emit time.
+fn consume_fence_body(lines: &[String], i: &mut usize, text: &mut String) {
+    let mut body: Vec<String> = Vec::new();
+    while *i < lines.len() {
+        let raw = &lines[*i];
+        let stripped = raw.trim();
+        *i += 1;
+        if stripped.starts_with("```") {
```
[the remark this answers is not in the document]
Comment (vmesel) [reply crates/outl-cli/src/cmd/import/normalize.rs:228]: Fixed in 6a588f1: `consume_fence_body` now tracks whether a closing fence was actually found and only re-emits it then. An unterminated fence (EOF first) is left without a synthesized closer, so malformed-but-existing input is no longer rewritten. Added regression test `unterminated_fence_keeps_no_synthetic_closer`.

File: crates/outl-actions/src/ingest.rs [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+/// `orphan_log` receives one line per orphan id surfaced during
+/// matching; pass `None` to suppress it.
+pub fn ingest_md_file(
+    workspace: &mut Workspace,
+    hlc: &HlcGenerator,
+    md_path: &Path,
+    orphan_log: Option<&Path>,
+) -> Result<ReconcileReport, ActionError> {
+    let slug = slug_from_path(md_path)
+        .ok_or_else(|| ActionError::InvalidSlug(md_path.display().to_string()))?;
```
[the remark this answers is not in the document]
Comment (vmesel) [reply crates/outl-actions/src/ingest.rs:58]: At this point `slug_from_path` returned `None` (empty or non-UTF8 file stem), so there is no derived slug to include yet — the path is the actionable info here. When a slug exists but fails validation, `open_or_create` already returns `InvalidSlug` carrying the slug. Leaving as-is.

File: crates/outl-actions/src/ingest.rs [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+
+/// Walk the whole tree, invoking `f` with each block's text.
+fn collect_refs<F: FnMut(&str)>(workspace: &Workspace, parent: NodeId, f: &mut F) {
+    for (id, _) in children_of(workspace, parent) {
+        if let Some(text) = workspace.block_text(id) {
+            f(&text);
+        }
+        collect_refs(workspace, id, f);
+    }
+}
```
[the remark this answers is not in the document]
Comment (vmesel) [reply crates/outl-actions/src/ingest.rs:242]: Realistic Logseq outlines are shallow (tens of levels, not thousands), so a stack overflow here is not a practical risk. The rest of the codebase also walks the tree recursively (backlinks, render projection), so converting just this one path to an explicit stack would be inconsistent without a real-world gain. Better handled as a codebase-wide policy change if we ever want one.

File: crates/outl-actions/src/ingest.rs [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+pub fn ingest_md_file(
+    workspace: &mut Workspace,
+    hlc: &HlcGenerator,
+    md_path: &Path,
+    orphan_log: Option<&Path>,
+) -> Result<ReconcileReport, ActionError> {
+    let slug = slug_from_path(md_path)
+        .ok_or_else(|| ActionError::InvalidSlug(md_path.display().to_string()))?;
+    let kind = kind_from_path(md_path);
+    let title = read_title(md_path).unwrap_or_else(|| slug.clone());
```
[the remark this answers is not in the document]
Comment (vmesel) [reply crates/outl-actions/src/ingest.rs]: Valid — `read_title` and `reconcile_md_with_page_id` each read+parse the file. It is not a bottleneck in practice (a full real-graph import runs in a couple of seconds), and deduping the work means threading the parsed contents through the `outl-md` reconcile API. Tracking as a follow-up rather than expanding this PR.

File: crates/outl-actions/src/ingest.rs [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+
+    // Create (or find) the page node under root with page-slug / page-kind.
+    open_or_create(workspace, hlc, &slug, &title, kind)?;
+
+    // Reconcile blocks under that exact node. The id must match the one
+    // open_or_create used, hence page_id_from_slug rather than letting
+    // reconcile mint a fresh random id when no sidecar exists yet.
+    let page_id = page_id_from_slug(&slug);
+    let report = reconcile_md_with_page_id(workspace, hlc, md_path, page_id, orphan_log)?;
```
[the remark this answers is not in the document]
Comment (vmesel) [reply crates/outl-actions/src/ingest.rs]: Same double-parse point as above — tracked as a follow-up.

**Review by avelino — CHANGES_REQUESTED** [review]
resolve conflict
I made a lot of changes to the lifecycle, breaking it into smaller packages to simplify maintenance/evolution, sync with main

**Review by avelino — COMMENTED** [review]
fix lands cleanly. importing a real Logseq graph as proof + the regression test for a pre-existing destination dir (the iCloud Ubiquity case) are exactly what was missing. veredict: comment / non-blocking, one edge case worth touching.

what stood out:

- migrated all 4 call sites (import, serve, TUI orphan scanner, mobile orphan scanner) to `outl_actions::ingest_md_file` in this same PR, instead of dropping a local helper into each client. that's exactly the reuse-first pattern the root `CLAUDE.md` prescribes, and it pays off as more clients land.
- `normalize.rs` is technically a new feature (not strictly the #43 fix), but bundling it makes sense, without normalization the page-creation fix still loses content on a real Logseq export. worth noting for later: when Roam / Bear / Obsidian importers land, this will grow into a collection, and at that point it may want to live under `outl-md` or a dedicated crate.

inline below covers the one edge case.

before merging, locally:

- `cargo fmt --check`
- `cargo clippy --workspace --all-targets -- -D warnings`
- `cargo test --workspace`
- `RUSTDOCFLAGS="-D warnings" cargo doc --workspace --no-deps`

File: crates/outl-actions/src/ingest.rs [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::page::list_all;
+    use outl_core::id::ActorId;
+    use std::fs;
+    use tempfile::TempDir;
+
+    fn workspace() -> (Workspace, HlcGenerator) {
+        let actor = ActorId::new();
```
Comment (avelino) [inline crates/outl-actions/src/ingest.rs:253]: heads up: `meta.title` flows straight into `format!("title:: {}\n", ...)` here with no sanitization. `extract_refs` walks raw block text, which can contain `\n` when a block has continuation lines (multi-paragraph body, or a `[[ref]]` split across lines). if the captured target is something like `[[Foo\nBar]]`, the stub `.md` ends up as:

```
title:: Foo
Bar
```

the parser then reads `title=Foo` plus an orphan block `Bar`, and re-importing the stub loses content silently.

cheap fix: `meta.title.replace('\n', ' ')` before the `format!`, or skip the stub entirely if the title contains a newline. either way, also worth a regression test alongside `implicit_pages_created_for_unresolved_refs`.

**Comment by avelino** [comment]
one architectural follow-up worth raising before this merges (non-blocking):

the workspace already has [`outl_actions::paste::normalize::normalize_external_syntax`](https://github.com/avelino/outl/blob/main/crates/outl-actions/src/paste/normalize.rs) doing **syntax-level** external-markdown coercion for the clipboard-paste path. it overlaps with this PR in three concrete spots:

| what `paste::normalize` already does | where it'd help in this PR |
|---|---|
| `\r\n` → `\n` | not handled anywhere in the PR; real Logseq exports from macOS sometimes carry CRLF |
| strips `id:: <ULID>` (with strict Crockford base32 validation, so 26 random alphanumeric chars don't get dropped by accident) | `crates/outl-cli/src/cmd/import/logseq.rs:159` reimplements this without the alphabet check |
| `[[June 2nd, 2026]]` → `[[2026-06-02]]`, `{{[[TODO]]}}` → `TODO`, `{{embed: ((blk-…))}}` → `!((blk-…))` | not handled; Logseq templates emit some of these forms too |

`import/normalize.rs` is still needed for **outline-level** restructuring (`## heading` → bullet, multi-paragraph merge, level clamping beyond 4→2, fence dedent/reindent) — `paste::normalize` doesn't touch block structure. so the two are complementary, not redundant.

the pipeline I'd suggest, as a follow-up PR if you'd rather keep this one focused:

```
file text
  → paste::normalize::normalize_external_syntax  (line endings, indent unit, tokens, dates, id::)
  → import::normalize::normalize_outline         (headings → bullets, multi-paragraph merge, clamp, fences)
  → parse → ingest_md_file
```

that way `logseq.rs::convert_file` drops the in-place `id::` and `((uid))` text replacement (the `((uid))` resolution still needs to live in the Logseq importer because it depends on the uid_index, but the rest moves up), and the importer is built out of shared, tested primitives instead of a parallel pass.

rationale is the reuse-first section of the root [`CLAUDE.md`](https://github.com/avelino/outl/blob/main/CLAUDE.md#reuse-first-no-parallel-implementations) — when there's already an upstream primitive doing the same job, the right move is to lift it (and generalize if needed) rather than fork. paying for that consolidation here also makes the future Roam / Bear / Obsidian importers cheaper: they all reuse the same paste-side normalization.

fine to land this PR as-is and tackle the consolidation in a follow-up. flagging it so it doesn't get lost.

**Comment by avelino** [comment]
@vmesel I'm going to close this PR, let's work on a more agnostic and extensible import module, so we can implement any import using the same core — unlike how it is today, where I rushed a quick implementation that ended up with a poor architecture
